Update nf-test guide with chapter (#3474)
nf-core/website: sites/docs/src/content/docs/tutorials/tests_and_test_data/components/03_testing_modules.md

│                                                                                                        │
│ Test Process BEDTOOLS_BAMTOBED                                                                         │
│                                                                                                        │
│   Test [824188d1] 'sarscov2 - bam' PASSED (3.277s)                                                     │
│   Test [f4f6429b] 'stub' PASSED (3.161s)                                                               │
│                                                                                                        │
│                                                                                                        │
│ SUCCESS: Executed 2 tests in 6.446s                                                                    │
│                                                                                                        │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────╯
INFO     All tests passed!
```

> ## Creating a new module with tests

Author: moved

    ]
    input[1] = 10
    """
  }
}
```

## Testing chained modules

For modules that depend on the output of other modules, use the `setup` method to define the dependency.

The `setup` method allows you to specify processes or workflows that need to be executed before the primary `when` block. It serves as a mechanism to prepare the required input data or set up essential steps prior to the primary processing block.

Within the setup block, you can use the `run` method to define and execute dependent processes or workflows.

### Syntax

The `run` method syntax for a process is as follows:

```groovy
run("ProcessName") {
    script "path/to/process/script.nf"
    process {
        // Define the process inputs here
    }
}
```

The `run` method syntax for a workflow is as follows:

```groovy
run("WorkflowName") {
    script "path/to/workflow/script.nf"
    workflow {
        // Define the workflow inputs here
    }
}
```

:::warning
Please keep in mind that changes in processes or workflows, which are executed in the setup method, can result in a failed test run.
:::

> ### Local `setup` method

Reviewer: Ah I see what you mean with global vs local

I would start with global and then give local. The description you've written makes more sense in this area.

your current description of the local one doesn't make it clear it's for a specific test, but if you say 'global is for all tests' then you can better infer 'local' means 'just a single test

> A `setup` method can be defined within a test case to execute a dependent process. This process generates input data required for the primary process. The `setup` block specifies the execution of the dependency, and the `when` block defines the processing logic for the module under test.

Reviewer: ```suggestion
A `setup` method can be defined within a test case to execute a dependent process. 
This process generates input data required for the primary process.
The `setup` block specifies the execution of the dependency, and the `when` block defines the processing logic for the module under test.
```

isn't this just repeating the intro section of setup block above?

Reviewer: (note I prefer this section!)

│ Load .nf-test/plugins/nft-anndata/0.1.0/nft-anndata-0.1.0.jar                                          │
│                                                                                                        │
│ Test Process ABRICATE_SUMMARY                                                                          │
│                                                                                                        │
│   Test [fc133477] 'Should run without failures' PASSED (11.658s)                                       │
│                                                                                                        │
│                                                                                                        │
│ SUCCESS: Executed 1 tests in 11.667s                                                                   │
│                                                                                                        │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────╯
INFO     All tests passed!
```

### Global `setup` method

> A global `setup` method can be defined for all tests within a `nextflow_process` definition. The `setup` is applied to multiple `test` cases, ensuring a consistent setup for each test. This approach is useful when multiple tests share the same setup requirements.

Reviewer: ```suggestion
A global `setup` method can be defined for all tests within a `nextflow_process` definition.
The `setup` is applied to multiple `test` cases, ensuring a consistent setup for each test.
This approach is useful when multiple tests share the same setup requirements.
```

Reviewer: Maybe it's worht adding an admoniton that it doesn't mean the setup process is only run once, and the output is shared across all tests, it just means you write less duplicated code - actually teh setup block is run for each test execution that requires it (a common misconception)

                    tuple([ id:'test2', single_end:false ],
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/haemophilus_influenzae/genome/genome.fna.gz', checkIfExists: true))
                ])
                """
            }
        }
    }

    test("first test") {
        when {
            process {
                """
                input[0] = ABRICATE_RUN.out.report.collect{ meta, report -> report }.map{ report -> [[ id: 'test_summary'], report]}
                """
            }
        }
        then {
            assert process.success
            assert snapshot(process.out).match()
        }
    }

    test("second test") {
        when {
            process {
                """
                input[0] = ABRICATE_RUN.out.report.collect{ meta, report -> report }.map{ report -> [[ id: 'test_summary'], report]}
                """
            }
        }
        then {
            assert process.success
            assert snapshot(process.out).match()
        }
    }
}
```

### Aliasing dependencies

> If you need to run the same process multiple times, you can set an `alias` for the process:

Reviewer: ```suggestion
If you need to run the same setup process multiple times for the same test but for different files, you can set an `alias` for the process:
```

```groovy
nextflow_process {

    // ...

    setup {

        run("UNTAR", alias: "UNTAR1") {
            script "modules/nf-core/untar/main.nf"
            process {
            """
            input[0] = Channel.fromList(...)
            """
            }
        }

        run("UNTAR", alias: "UNTAR2") {
            script "modules/nf-core/untar/main.nf"
            process {
            """
            input[0] = Channel.fromList(...)
            """
            }
        }

        run("UNTAR", alias: "UNTAR3") {
            script "modules/nf-core/untar/main.nf"
            process {
            """
            input[0] = Channel.fromList(...)
            """
            }
        }
    }

    test("Test with three different inputs") {
        when {
            process {
                """
                input[0] = UNTAR1.out.untar.map{ it[1] }
                input[1] = UNTAR2.out.untar.map{ it[1] }
                input[2] = UNTAR3.out.untar.map{ it[1] }
                """
            }
        }

        then {
            // ...
        }
    }
}
```

## Updating module snapshots

> When module outputs change (e.g., due to version bumps), you need to update snapshots:

Reviewer: ```suggestion
Whenever a module is updated resulting in changes to output (e.g., due to version bumps of the tool itself), you will need to update snapshots:
```

│ https://www.nf-test.com                                                                                │
│ (c) 2021 - 2024 Lukas Forer and Sebastian Schoenherr                                                   │
│                                                                                                        │
│ Load .nf-test/plugins/nft-bam/0.5.0/nft-bam-0.5.0.jar                                                  │
│ Load .nf-test/plugins/nft-compress/0.1.0/nft-compress-0.1.0.jar                                        │
│ Load .nf-test/plugins/nft-vcf/1.0.7/nft-vcf-1.0.7.jar                                                  │
│ Load .nf-test/plugins/nft-csv/0.1.0/nft-csv-0.1.0.jar                                                  │
│ Load .nf-test/plugins/nft-utils/0.0.3/nft-utils-0.0.3.jar                                              │
│ Load .nf-test/plugins/nft-fastq/0.0.1/nft-fastq-0.0.1.jar                                              │
│ Load .nf-test/plugins/nft-anndata/0.1.0/nft-anndata-0.1.0.jar                                          │
│ Warning: every snapshot that fails during this test run is re-record.
```

> Once the test passes, the snapshot will be updated and the test re-run to verify the snapshot is stable.

Reviewer: ```suggestion
If you are using nf-core tools for executing the tests, once the test passes, the snapshot will be updated and the test(s) will be re-run for you to verify the snapshot is stable.

If it is not stable, you will need to refine your assertions.
```

## Testing with nf-test

> You can also run the tests with nf-test directly using the `--tag` option:

Reviewer: ```suggestion
If you are running with nf-test directly, you can use the `--tag` option to specify which module within the repository to test:
```

---
title: "Testing nf-core/modules"
subtitle: Writing comprehensive tests for nf-core modules with nf-test
weight: 30
---

## Overview

The nf-test framework in the nf-core ecosystem enables comprehensive testing for processes, workflows, and pipelines.

This chapter covers the fundamentals of nf-core module testing, from basic syntax to advanced scenarios involving chained modules.

## Module structure

Before diving into testing, let's understand the typical structure of a nf-core module:

```
modules/nf-core/tool/subtool/
├── main.nf              # Process definition
├── meta.yml             # Module metadata
└── tests/               # Testing directory
    ├── main.nf.test     # Test definitions
    └── nextflow.config  # Optional: test-specific config
```

### Basic test syntax

The basic syntax for a process test follows this structure:

```groovy title="main.nf.test"
nextflow_process {
    name "<NAME>"
    script "<PATH/TO/NEXTFLOW_SCRIPT.nf>"
    process "<PROCESS_NAME>"

    test("<TEST_NAME>") {
        when {
        }

        then {
        }
    }
}
```

:::note
**Key points:**

> - Script paths starting with `./` or `../` are relative to the test script's location.

Reviewer: OK it seems rather unnecessary detail in a very broad/generic overview? I would remove it personally

---
title: "Testing nf-core/modules"
subtitle: Writing comprehensive tests for nf-core modules with nf-test
weight: 30
---

## Overview

The nf-test framework in the nf-core ecosystem enables comprehensive testing for processes, workflows, and pipelines.

This chapter covers the fundamentals of nf-core module testing, from basic syntax to advanced scenarios involving chained modules.

## Module structure

Before diving into testing, let's understand the typical structure of a nf-core module:

> ```

Reviewer: ```suggestion
```tree
```

Within the `setup` block, you can use the `run` method to define and execute multiple dependent processes or workflows.

Here's a basic example of how a setup block looks:

```groovy
nextflow_process {
    name "Test Process MY_MODULE"
    script "../main.nf"
    process "MY_MODULE"

    test("test with setup") {
        setup {
            run("UPSTREAM_MODULE") {
                script "../../upstream/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test' ],
                        file('path/to/input.txt')
                    ]
                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = UPSTREAM_MODULE.out.my_output
                """
            }
        }

        then {
            assert process.success
        }
    }
}
```

:::warning
Please keep in mind that changes in processes or workflows, which are executed in the setup method, can result in a failed test run.
:::

### Global `setup` method (for all tests)

> A global `setup` method can be defined for all tests within a `nextflow_process` definition.

Reviewer: Ah sorry, I misunderstood - my brain apparently never actually registered it as it's been in every single module 😆

            process {
            """
            input[0] = Channel.fromList(...)
            """
            }
        }
    }

    test("Test with three different inputs") {
        when {
            process {
                """
                input[0] = UNTAR1.out.untar.map{ it[1] }
                input[1] = UNTAR2.out.untar.map{ it[1] }
                input[2] = UNTAR3.out.untar.map{ it[1] }
                """
            }
        }

        then {
            // ...
        }
    }
}
```

### Testing parameter variations

Some modules may require additional parameters added to the test command to successfully run.

These can be specified using a params input and an `ext.args` variable within the process scope of the `nextflow.config` file, which exists alongside the test files themselves (and is automatically loaded when the test workflow `main.nf` is executed).

If your module requires a `nextflow.config` file to run, create the file to the module’s `tests/` directory and add the following code to use parameters defined in the `when` scope of the test.

```bash
touch modules/nf-core/<tool>/<subtool>/tests/nextflow.config
```

```nextflow.config

> 

Reviewer: ```suggestion
```

process {
  withName: 'MODULE' {
    ext.args = params.module_args
  }
}
```

You do not need to modify the contents of this file any further.

> Then add the config to the main.nf.test file and supply the params in the when section of the test.

Reviewer: ```suggestion
Then add the config to the `main.nf.test` file and supply the params in the `when` section of the test.
```

```main.nf.test

> 

Reviewer: ```suggestion
```

│ Load .nf-test/plugins/nft-fastq/0.0.1/nft-fastq-0.0.1.jar                                              │
│ Load .nf-test/plugins/nft-anndata/0.1.0/nft-anndata-0.1.0.jar                                          │
│ Warning: every snapshot that fails during this test run is re-record.
```

If you are using nf-core tools for executing the tests, once the test passes, the snapshot will be updated and the test(s) will be re-run for you to verify the snapshot is stable.

If it is not stable, you will need to refine your assertions.

## Testing with nf-test

If you are running with nf-test directly, you can use the `--tag` option to specify which module within the repository to test:

```bash
nf-test test --profile docker --tag abricate/summary

# or specify test path
nf-test test --profile docker modules/nf-core/abricate/summary/tests/main.nf.test

# update snapshots
nf-test test --profile docker modules/nf-core/abricate/summary/tests/main.nf.test --update-snapshot
```

This will run the tests for the module and display the results, including any failures or snapshot mismatches.

```bash

> 🚀 nf-test 0.9.0

Reviewer: Again this seems somewhat unnecesasry

---
title: "Testing nf-core/modules"
subtitle: Writing comprehensive tests for nf-core modules with nf-test
weight: 30
---

## Overview

The nf-test framework in the nf-core ecosystem enables comprehensive testing for processes, workflows, and pipelines.

This chapter covers the fundamentals of nf-core module testing, from basic syntax to advanced scenarios involving chained modules.

> ### Basic test syntax

Reviewer: Actually in retrospect, I think this basic test syntax section can be removed or moved to the workflow example.

The 'understanding snapshots' is a great lead in for someone to understand the concept behind nf-test and CI testing in general, and what the tutorial will give you by the end. The section after is how to to write the tests that allow you to achieve the snapshots - if you see the 'story' I'm going ofr here.?

Reviewer: As in 'understanding snapshots' is describing the final goal/aim of the tutorial

The basic syntax for a module test that we define in `main.nf.test` follows this structure:

```groovy title="main.nf.test"
nextflow_process {
    name "<NAME>"
    script "<PATH/TO/NEXTFLOW_SCRIPT.nf>"
    process "<PROCESS_NAME>"

    test("<TEST_NAME>") {
        when {
        }

        then {
        }
    }
}
```

**Components:**

- `nextflow_process` - Main test block for the process
- `name` - Descriptive test suite identifier (e.g., "Test Process FASTQC")
- `script` - Path to the Nextflow script (usually `"../main.nf"`)
- `process` - Exact process name from the script (case-sensitive)
- `test()` - Individual test case with descriptive name
- `when` - Test input setup and execution parameters
- `then` - Assertions to verify expected behavior and outputs

:::note
**Key points:**

- Script paths starting with `./` or `../` are relative to the test script's location.
  :::

## Understanding snapshots

**What are snapshots?**

> Snapshots are nf-test's way of capturing and validating the expected outputs of your tests. When you run a test for the first time, nf-test creates a `.snap` file containing checksums, file names, and other metadata about the outputs.

Reviewer: ```suggestion
Snapshots are nf-test's way of capturing and validating the expected outputs of your tests. When you run a test for the first time, nf-test creates a `.snap` file containing checksums, file names, and other metadata about that describe how a module's output should look like.
```

```groovy
process CAT {
    input:
    tuple val(meta), path(file_in)

    output:
    tuple val(meta), path("${prefix}"), emit: file_out
    path "versions.yml"               , emit: versions

    script:
    prefix = "${meta.id}.txt"
    """
    cat ${file_in} > ${prefix}
    """
}
```

This process simply concatenates input files and outputs a single file.

### Step 2: Write the test

Create `modules/cat/tests/main.nf.test`:

```groovy
nextflow_process {
    name "Test Process CAT"
    script "../main.nf"
    process "CAT"

    test("simple file concatenation") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map with sample identifier
                    file('test_file.txt')  // input file
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },           // Check process completed successfully
                { assert snapshot(process.out).match() } // Validate outputs match expected snapshot
            )
        }
    }
}
```

**Key components explained:**

> 

Reviewer: I would  actually describe the top level name/script/process/`when/test` scopes are for  and then finally  move into the specific detail of the id/file/process.success/snapshot

- `[ id:'test' ]` - **meta map**: Contains sample metadata (ID, conditions, etc.)
- `file('test_file.txt')` - **input file**: The file to process
- `process.success` - **success check**: Ensures the process didn't fail
- `snapshot(process.out).match()` - **output validation**: Compares all outputs to stored snapshot

### Step 3: Run the test

> 

Reviewer: ```suggestion

Once you've written the test, you can generate the initial snapshot by running the test.

```

```bash
cd path/to/modules
nf-core modules test cat --profile docker (EXAMPLE)
```

**What happens:**

1. nf-test creates a temporary work directory `.nf-test/`
2. Runs the CAT process with your test inputs
3. **First run**: Generates `tests/main.nf.test.snap` with output checksums
4. **Second run**: Validates outputs match the generated snapshot
5. Reports PASS/FAIL

### Step 4: Examine the generated snapshot

After running, check `tests/main.nf.test.snap`:

```json
{
  "simple file concatenation": {
    "content": [
      {
        "0": [
          // First output channel (file_out)
          [{ "id": "test" }, "test.txt:md5,d41d8cd98f00b204e9800998ecf8427e"]
        ],
        "1": [
          // Second output channel (versions)
          "versions.yml:md5,c50aa59475ab20752531545a585f8f2d"
        ]
      }
    ]
  }
}
```

> Now you understand the full test cycle! The snapshot ensures your module produces consistent, expected outputs every time.

Reviewer: ```suggestion

You can see the the outputs of the module, with an associated `md5sum` which concisely describes the exact contents of the file.

Now you understand the full test cycle! The snapshot ensures your module produces consistent, expected outputs every time. 

Any time you change the module - you can run the test again to check to see you did not unexpectedly change an output that should not have been changed.
```

### Essential Assertions

Tests then use assertions to verify the expected output of the process specified in the `then` block.

You can specify multiple assertions to be evaluated together in a single test by specifying them within an `assertAll` block.

Nextflow process output channels that lack explicit names (i.e., when no `meta` map is present) can be addressed using square brackets and the corresponding index, for example `process.out[0]` for the first channel and `process.out[0][1]` for the second element of the first channel.

```groovy
// Process completion status
assert process.success
assert process.exitStatus == 0

// Output channels
assert process.out.my_channel != null
assert process.out.my_channel.size() == 3
assert process.out.my_channel.get(0) == "expected_value"

// For unnamed channels, use index notation
assert process.out[0] != null
assert process.out[0].size() == 3

// Group assertions to see all failures at once
assertAll(
    { assert process.success },
    { assert snapshot(process.out).match() }
)
```

:::note
For more nf-test assertion patterns, see the [nf-test assertions examples documentation](./07_assertions.md).
:::

## Testing an existing module

> Let's examine testing the `bedtools/bamtobed` module, which is a simple standalone module:

Reviewer: ```suggestion
Let's now examine how we can test an existing module, such as  `bedtools/bamtobed` that is a simple standalone module:
```

│ Test Process BEDTOOLS_BAMTOBED                                                                         │
│                                                                                                        │
│   Test [824188d1] 'sarscov2 - bam' PASSED (3.277s)                                                     │
│   Test [f4f6429b] 'stub' PASSED (3.161s)                                                               │
│                                                                                                        │
│                                                                                                        │
│ SUCCESS: Executed 2 tests in 6.446s                                                                    │
│                                                                                                        │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────╯
INFO     All tests passed!
```

### Stub mode

All nf-core modules require a `stub:` section to allow 'dry run'-like functionality for pipelines.

> This is something that we always want to test for.

Reviewer: ```suggestion
This is something that we always want to test for, even if your module also has tests with real data.
```

Furthermore, in some cases the module will produce output data that is too big for GitHub actions nodes, or the tests take too long or require too much resources.
Therefore, the only option is to test the module in `stub` mode.

Therefore when adding a stub test, you need to tell a test to run in stub mode by adding the `-stub` option as follows:

```main.nf.test
process "MODULE"
config "./nextflow.config"

test("custom evalue") {
  options "-stub"
    when {
        process {
            // test implementation
        }
    }
}
```

:::note
For modules that can never run full tests due to data being too large or requiring too much resources, this option can alternatively be added at the top of `main.nf.test` to have all tests run in stub mode.
:::

## Testing chained modules

> In some cases, rather than directly linking to pre-made test-data files, it may make sense to run an 'upstream module' in your test to output the required inputs of the module you want to test.

Reviewer: ```suggestion
In some cases, rather than directly linking to pre-made test-data files within the `when:` scope, it may make sense to run an 'upstream module' in your test that produces the required inputs of the module you want to test.
```

    }

    test("Test with three different inputs") {
        when {
            process {
                """
                input[0] = UNTAR1.out.untar.map{ it[1] }
                input[1] = UNTAR2.out.untar.map{ it[1] }
                input[2] = UNTAR3.out.untar.map{ it[1] }
                """
            }
        }

        then {
            // ...
        }
    }
}
```

### Testing parameter variations

Some modules may require additional parameters added to the test command to successfully run.

These can be specified using a `params` input and an `ext.args` variable within the process scope of the `nextflow.config` file, which exists alongside the test files themselves (and is automatically loaded when the test workflow `main.nf` is executed).

If your module requires a `nextflow.config` file to run, create the file to the module’s `tests/` directory and add the following code to use parameters defined in the `when` scope of the test.

```bash
touch modules/nf-core/<tool>/<subtool>/tests/nextflow.config
```

```nextflow.config
process {
  withName: 'MODULE' {
    ext.args = params.module_args
  }
}
```

You do not need to modify the contents of this file any further.

> Then add the config to the `main.nf.test` file and supply the params in the `when` section of the test.

Reviewer: ```suggestion
Then import the config to the `main.nf.test` file and supply the params in the `when` section of the test.
```

    ]
    """
  }
}
```

### Choosing Parameter Configuration Methods

**Use `nextflow.config` when:**

- Multiple tests in the same module need the same parameter structure
- You need complex process-specific configurations (memory, cpus, etc.)
- Parameters require process selectors or conditional logic
- You want to maintain consistent configuration across all tests

**Use inline `params` blocks when:**

- Testing different parameter values in individual tests
- You need test-specific parameter overrides
- Parameters are simple and don't require process configuration
- You want to keep test parameters close to the test logic

**Example comparison:**

```groovy
// nextflow.config approach - good for consistent configuration
process {
  withName: 'BLAST_BLASTN' {
    ext.args = params.blast_args
    memory = params.blast_memory ?: '8.GB'
  }
}

// Inline params approach - good for test-specific values
test("custom evalue") {
    when {
        params {
            blast_args = '-evalue 0.001 -max_target_seqs 10'
        }
        process {
            // test implementation
        }
    }
}
```

> 

Reviewer: ```suggestion
```

O think you can delete this comparison as you don't introduce the config only method?

## Updating module snapshots

> Whenever a module is updated resulting in changes to output (e.g., due to version bumps of the tool itself), you will need to update snapshots:

Reviewer: ```suggestion
Whenever a module is updated that results in changes to output (e.g., due to version bumps of the tool itself), you will need to update snapshots:
```
